### ISSUE TITLE
Incorrect generic inferred for function inside object spread

### ISSUE THREAD

**Issue by OliverJAsh** [microsoft/TypeScript#53016]
# Bug Report

### 🔎 Search Terms

- generic function
- spread operator
- intersection

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about generics

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3gBwJbZAC54AeAQQBp4AhAPgAopjL4oAjYx5gSngF5aNXgJoBuAFDiMATwLwACjCTY0-eAG9E0NGgByUALZF4mGLhQBzeAF8J40JFgJk6LFGDAAwtr2HjJABV4EAAPDBAUYFVFZTQGRx19I2JTcwthQXo0YgCM+AC7B2g4RFRMHCUVYhiVOxdyuDQAVwgMasrVADINeDZYFIwzSxs1dXF4eAB6ACpp8Ynp+EAZcngAVRRQgjBw4GCYJRhCefhFshgLJqMULCQAM3gZOQByGrQn+FxVFCQ3HVwLFDsCAIDBIHCwPzhGDwO4PWQIF4deDdTR9Q4mQZpUQ2J4AOmOiwmNRAMBk8CeaPen3gBk+aDSHxQcOer3ebCaWDgAEcmrg4LtzMyEaj+hihhZsdY8QTJvNcfK8AR6NgOlR3F4fEkQPQAES3JBIHXcbgUSQTNHEHUARgATABmHWm2xAA)

### 💻 Code

```ts
declare const pipe: <A, B>(a: A, ab: (a: A) => B) => B;

type Props = { className: string };

declare const addClassName: <T extends Props>(className: string) => (s: T) => T;

declare const props: Props;

const result: Props & { bar: string } = {
  /**
   * ❌ Unexpected error:
   * Argument of type 'Props' is not assignable to parameter of type 'Props & { bar: string; }'.
   *   Property 'bar' is missing in type 'Props' but required in type '{ bar: string; }'.
   *
   * The generic of `addClassName` is inferred as `Props & { bar: string; }`
   */
  ...pipe(props, addClassName("foo")),

  bar: "123",
};
```

### 🙁 Actual behavior

- Unexpected error (see code comments above).
- The generic of `addClassName` is inferred as `Props & { bar: string; }`.

### 🙂 Expected behavior

- No error.
- The generic of `addClassName` should be inferred as `Props`, based on the value we are piping into the function (`props`).

---

Known workarounds:

```ts
/**
 * Workaround 1: extract spreaded value to a variable
 */
const x = pipe(props, addClassName("foo"));
const result2: Props & { bar: string } = {
  ...x,

  bar: "123",
};

/**
 * Workaround 1: explicit generic type
 */
const result3: Props & { bar: string } = {
  ...pipe(props, addClassName<Props>("foo")),

  bar: "123",
};

```

**Comment by RyanCavanaugh** [comment]
Given the form `const x: T = { ...e, p: 0 }` it's not apparent if the "correct" contextual type to give `e` is `T` or `Omit<T, 'p'>`, and depending on the intent of the code it could definitely be either. It depends on whether `p` is an "intentional overwrite" of something in `e` or not, and there's no way to know.

That said, the contextual inference is definitely spoiling the pot. It seems like a better definition of `pipe` would be
```ts
declare const pipe: <A, B extends (a: A) => unknown>(a: A, ab: B) => ReturnType<B>;
```
